### PR TITLE
Delay rpc calls until node synced

### LINKED ISSUE #558: starting up the rpc client until after the initial sync is completed.
### What's wrong

The RPC client shouldn't be serving before the node is fully synced.

### How can we fix it

Should return HTTP 503

TODO: Need to figure out how to implement this

### PR DESCRIPTION
**Pull request by kautukkundan**
closes #558

### CODE REVIEW

File: test/client/integration.test.ts [hunk 1]
```diff
@@ -160,5 +160,5 @@ describe("Client Integration", function() {
 
         assert.equal(storageSyncer.transactions.count(), maxTransfers);
         assert.equal(storagePacker.transactions.count(), maxTransfers);
-    }).timeout(600000);
+    }).timeout(900000);
```
Comment (kautukkundan) [inline test/client/integration.test.ts:163]: increased time limit as tests keep failing at the edge of 10min

Comment (jacque006) [inline test/client/integration.test.ts:163]: Why was this needed?

File: ts/client/index.ts [hunk 2]
```diff
@@ -15,6 +16,16 @@ async function main() {
...
 
+    fast.addHook("onRequest", async (request, reply) => {
+        if (node.getSyncMode() === SyncMode.INITIAL_SYNCING) {
+            return reply.status(503).send({
+                message: "Initial sync incomplete",
+                error: "RPC unavailable",
+                statusCode: 503
+            });
+        }
+    });
```
Comment (jacque006) [inline ts/client/index.ts:27]: I think we can move this into `node.ts` and remove the getter for syncMode, more appropriate for node to manage this hook.

I addition, one downside of this is that all endpoints will return this status including our process health check (`/health`) which would be used by container orchestration platforms and uptime checks in general to make sure the node is still responding to pings. We should have the hook ignore `/health` or consider adding the hook only to the endpoints registered in `RPC.ts`.

Comment (kautukkundan) [reply ts/client/index.ts:27]: the fastify Instance is only present inside the "static" init method of the node class, adding the onRequest hook inside that function won't work at runtime as we need to check the syncMode at every API call. to make node manage the hook, fastify instance must be saved as a class variable and then the hook must be added inside the `start` method.

Comment (jacque006) [reply ts/client/index.ts:27]: You can move to node without saving instance since the `syncer` service is in scope:
Screenshot from 2021-07-22 17-41-53

Comment (kautukkundan) [reply ts/client/index.ts:27]: oh so sorry, my bad
i was thinking that the `syncer` instance gets destroyed outside the scope of this init function. thanks 😃

**Comment by kautukkundan** [comment]
Some notes -

1. Based on Jake's previous comment
> I addition, one downside of this is that all endpoints will return this status including our process health check (/health) which would be used by container orchestration platforms and uptime checks in general to make sure the node is still responding to pings. We should have the hook ignore /health or consider adding the hook only to the endpoints registered in RPC.ts.

I guess this is sorted as the fastify hook only applies to the fastify endpoints. since the `/health` endpoint is created with the terminus, it does not have the `onRequest` hook applied to it. I have manually tested this as well. 

2. I tried to write a test case for testing out the rpc endpoints. However, the server requires geth node to be running already on port 8545 with the contracts deployed and registered. This needs some discussion if automated testing is required for the api right now. Also should that be included in the existing test suite or a separate test suite specifically for testing out APIs which runs separately after the default test suite.

File: ts/client/node.ts [hunk 3]
```diff
@@ -89,9 +89,18 @@ export class HubbleNode {
             throw new Error("watcher is currently not supported");
         }
 
-        // In the future, we will want to delay starting up the rpc client
-        // until after the initial sync is completed (HTTP 503).
-        // https://github.com/thehubbleproject/hubble-contracts/issues/558
+        fast.addHook("onRequest", async (_request, reply) => {
+            console.log("hook ran");
+
```
Comment (jacque006) [inline ts/client/node.ts:94]: Remove

**Review by jacque006 — APPROVED** [review]
Asides from errant console.log, looks good

File: ts/client/node.ts (hunk 3 above)
Comment (kautukkundan) [reply ts/client/node.ts:94]: oops! missed this

**Comment by jacque006** [comment]
For testing, given the amount or work it will involve to setup I think we can skip for now. Manual testing is good :+1:
